### PR TITLE
[Core] dlmalloc allocate bottom-most memory chunk failed 

### LINKED ISSUE #21310: [Bug] dlmalloc allocate bottom-most memory chunk failed
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

I find `dlmalloc` can't allocate the bottom-most memory chunk which is just freed by deleting the same memory size RayObject:

It can reproduce by this python code:
```python
import ray
import numpy as np
ray.init(
    object_store_memory=300 * 1024**2,
    _system_config={
        "max_io_workers": 4,
        "automatic_object_spilling_enabled": True,
    })
res = []
data = np.random.randint(
    low=0, high=256, size=(90 * 1024**2, ), dtype=np.uint8)
for _ in range(3):
    res.append(ray.put(data))
# keep reference for second and third object, force evict first object
_ = ray.get(res[1:])
ray.put(data)
```

then you can see this from `raylet.out`:
```bash
[2021-12-31 17:06:01,438 D 47800 2246939] object_lifecycle_manager.cc:219: evicting object ffffffffffffffffffffffffffffffffffffffff0100000001000000
[2021-12-31 17:06:01,438 D 47800 2246939] plasma_allocator.cc:114: deallocating 94372093 at 0x115726040
[2021-12-31 17:06:01,438 D 47800 2246939] object_store.cc:28: attempting to create object ffffffffffffffffffffffffffffffffffffffff0100000004000000 size 94372087
[2021-12-31 17:06:01,438 D 47800 2246939] plasma_allocator.cc:81: allocating 94372093
[2021-12-31 17:06:01,438 D 47800 2246939] dlmalloc.cc:224: fake_mmap called once already, refusing to overcommit: 94633984
[2021-12-31 17:06:01,438 D 47800 2246939] plasma_allocator.cc:83: allocated 94372093 at 0x0
```
`dlmalloc` has already deallocated 94372093 bytes at 0x115726040, but still can't allocate the same size at the bottom-most memory chunk, until `FallbackAllocate` been called:

```bash
[2021-12-31 17:06:03,441 D 47800 2246939] plasma_allocator.cc:94: fallback allocating 94372093
[2021-12-31 17:06:03,441 I 47800 2246939] dlmalloc.cc:146: create_and_mmap_buffer(94375944, /tmp/ray/plasmaXXXXXX)
[2021-12-31 17:06:03,442 D 47800 2246939] dlmalloc.cc:250: 0x10c24a008 = fake_mmap(94375944)
[2021-12-31 17:06:03,442 D 47800 2246939] plasma_allocator.cc:96: allocated 94372093 at 0x10c24a040
[2021-12-31 17:06:03,442 D 47800 2246939] object_store.cc:50: create object ffffffffffffffffffffffffffffffffffffffff0100000004000000 succeeded
```

I print memory chunk:
```bash
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:118: mallinfo , arena: 402653176, ordblks: 4, smblks: 0, hblks: 0, hblkhd: 8, usmblks: 402653184, fsmblks: 0, uordblks: 360000824, fordblks: 42652360, keepcost: 42652152,
allocated: 360000777, footprint_limit: 402653184
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:131: ------------------------- mallinfo -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:132: ------------------------- malstats -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:133: malstats , dvsize: 0, dvpointer: 0, topsize: 42652152, toppointer0x7f6729751dc0
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:136: ------------------------- blockinfo -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:149: ------------------------- seginfo 0 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:154: ------------------------- chunkinfo 0 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:161: chunk pointer: 0x7f6713fff010, chunk in use: 0, chunk size: 32
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:165: ------------------------- chunkinfo 0 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:154: ------------------------- chunkinfo 1 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:161: chunk pointer: 0x7f6713fff030, chunk in use: 1, chunk size: 120000272
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:165: ------------------------- chunkinfo 1 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:154: ------------------------- chunkinfo 2 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:161: chunk pointer: 0x7f671b26ff40, chunk in use: 0, chunk size: 48
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:165: ------------------------- chunkinfo 2 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:154: ------------------------- chunkinfo 3 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:161: chunk pointer: 0x7f671b26ff70, chunk in use: 1, chunk size: 120000272
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:165: ------------------------- chunkinfo 3 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:154: ------------------------- chunkinfo 4 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:161: chunk pointer: 0x7f67224e0e80, chunk in use: 0, chunk size: 48
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:165: ------------------------- chunkinfo 4 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:154: ------------------------- chunkinfo 5 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:161: chunk pointer: 0x7f67224e0eb0, chunk in use: 1, chunk size: 120000272
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:165: ------------------------- chunkinfo 5 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:169: ------------------------- seginfo 0 -------------------------
[2021-12-27 18:00:35,598 D 110507 110524] plasma_allocator.cc:179: ------------------------- blockinfo -------------------------
```
There seems to be a lot of memory fragmentation, I think this is caused by the difference between `dlmalloc` MALLOC_ALIGNMENT(16-bytes in my environment) and the arg alignment (64-bytes) which send to function `dlmemalign`. 

Then I changed MALLOC_ALIGNMENT to 64-bytes in `dlmalloc.cc`, I find the bottom-most memory chunk can be allocated, and memory fragmentation is also reduced:

```bash
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:118: mallinfo , arena: 402653144, ordblks: 1, smblks: 0, hblks: 0, hblkhd: 40, usmblks: 402653184, fsmblks: 0, uordblks: 360001000, fordblks: 42652184, keepcost: 42652008,
allocated: 360000777, footprint_limit: 402653184
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:131: ------------------------- mallinfo -------------------------
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:132: ------------------------- malstats -------------------------
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:133: malstats , dvsize: 0, dvpointer: 0, topsize: 42652008, toppointer0x7f2241751df0
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:136: ------------------------- blockinfo -------------------------
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:149: ------------------------- seginfo 0 -------------------------
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:154: ------------------------- chunkinfo 0 -------------------------
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:161: chunk pointer: 0x7f222bfff030, chunk in use: 1, chunk size: 120000320
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:165: ------------------------- chunkinfo 0 -------------------------
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:154: ------------------------- chunkinfo 1 -------------------------
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:161: chunk pointer: 0x7f223326ff70, chunk in use: 1, chunk size: 120000320
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:165: ------------------------- chunkinfo 1 -------------------------
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:154: ------------------------- chunkinfo 2 -------------------------
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:161: chunk pointer: 0x7f223a4e0eb0, chunk in use: 1, chunk size: 120000320
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:165: ------------------------- chunkinfo 2 -------------------------
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:169: ------------------------- seginfo 0 -------------------------
[2021-12-27 18:28:16,326 D 5223 5240] plasma_allocator.cc:179: ------------------------- blockinfo -------------------------
```

### Versions / Dependencies

OS: mac 11.4
python: 3.7.9
ray: 1.9.1

### Reproduction script

It can reproduce by this python code:
```python
import ray
import numpy as np
ray.init(
    object_store_memory=300 * 1024**2,
    _system_config={
        "max_io_workers": 4,
        "automatic_object_spilling_enabled": True,
    })
res = []
data = np.random.randint(
    low=0, high=256, size=(90 * 1024**2, ), dtype=np.uint8)
for _ in range(3):
    res.append(ray.put(data))
# keep reference for second and third object, force evict first object
_ = ray.get(res[1:])
ray.put(data)
```

### Anything else

I don’t know if there is a better way than modifying MALLOC_ALIGNMENT, do you guys have any suggestions?

### Are you willing to submit a PR?

- [x] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by Catch-Bull**
## Why are these changes needed?

fix dlmalloc allocate bug,  details in here #21310

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jjyao** [comment]
@Catch-Bull The issue you linked is a Ray Tune issue, is it the correct one?

File: python/ray/tests/test_plasma_unlimited.py [hunk 1]
```diff
@@ -247,6 +248,34 @@ def test_fallback_allocation_failure(shutdown_only):
...
+    data = np.random.randint(
+        low=0, high=256, size=(90 * 1024**2, ), dtype=np.uint8)
+    for _ in range(3):
+        res.append(ray.put(data))
+    # keep reference for second and third object, force evict first object
+    _ = ray.get(res[1:])  # noqa
+    # keep reference for fourth object, avoid released by plasma GC.
+    __ = ray.put(data)  # noqa
+
+    # Check fourth object whether allocate to disk
```
Comment (ericl) [inline python/ray/tests/test_plasma_unlimited.py:272]: Can you use `check_spilled_mb` to see if it was disk allocated instead?

File: src/ray/object_manager/plasma/dlmalloc.cc [hunk 2]
```diff
@@ -56,6 +56,9 @@ int fake_munmap(void *, int64_t);
 #define HAVE_MORECORE 0
 #define DEFAULT_MMAP_THRESHOLD MAX_SIZE_T
 #define DEFAULT_GRANULARITY ((size_t)128U * 1024U)
+// Copied from plasma_allocator.cc variable kAllocationAlignment,
+// make sure to keep in sync
```
Comment (ericl) [inline src/ray/object_manager/plasma/dlmalloc.cc:60]: Can you add a link to GH issue for context?

File: python/ray/tests/test_plasma_unlimited.py (hunk 1 above)
Comment (Catch-Bull) [reply python/ray/tests/test_plasma_unlimited.py]: done.

File: src/ray/object_manager/plasma/dlmalloc.cc (hunk 2 above)
Comment (Catch-Bull) [reply src/ray/object_manager/plasma/dlmalloc.cc]: done.

**Comment by Catch-Bull** [comment]
> @Catch-Bull The issue you linked is a Ray Tune issue, is it the correct one?

oh... it's my mistake, I've made it correct.

File: python/ray/tests/test_plasma_unlimited.py [hunk 3]
```diff
@@ -247,6 +248,31 @@ def test_fallback_allocation_failure(shutdown_only):
...
+        low=0, high=256, size=(90 * 1024**2, ), dtype=np.uint8)
+    for _ in range(3):
+        res.append(ray.put(data))
+    # keep reference for second and third object, force evict first object
+    _ = ray.get(res[1:])  # noqa
+    # keep reference for fourth object, avoid released by plasma GC.
+    __ = ray.put(data)  # noqa
+
+    # Check fourth object allocate in memory.
+    _check_spilled_mb(address, spilled=180)
```
Comment (scv119) [inline python/ray/tests/test_plasma_unlimited.py:273]: you can also write a unit test in C++ demonstrate the issue. Here are some examples you can follow:
https://github.com/ray-project/ray/blob/master/src/ray/object_manager/plasma/test/fallback_allocator_test.cc#L35

Comment (Catch-Bull) [reply python/ray/tests/test_plasma_unlimited.py:272]: done.

**Comment by scv119** [comment]
Thanks for addressing the comments. Looks the lint issues is unrelated but let me retry them before merging.

**Review by scv119 — APPROVED** [review]

**Review by scv119 — APPROVED** [review]

**Pull request merged**